### PR TITLE
types,expression: Fix parse time from string

### PR DESCRIPTION
**Pull request by spongedu**
### What is changed and how it works?
Currently TiDB fail to deal with some cases when parse `string` to `datetime`, This pr fixes these cases. For example, in MySQL:
```sql
mysql> select cast("1701020301" as datetime);
+--------------------------------+
| cast("1701020301" as datetime) |
+--------------------------------+
| 2017-01-02 03:01:00            |
+--------------------------------+
1 row in set (0.00 sec)

mysql> select cast("170102034" as datetime);
+-------------------------------+
| cast("170102034" as datetime) |
+-------------------------------+
| 2017-01-02 03:04:00           |
+-------------------------------+
1 row in set (0.00 sec)

```

While in TiDB:
```sql
tidb> select cast("1701020301." as datetime);
ERROR 1105 (HY000): invalid time format
tidb> select cast("1701020301" as datetime);
ERROR 1105 (HY000): invalid time format
tidb> select cast("170102034" as datetime);
ERROR 1105 (HY000): invalid time format
```

### Check List 

Tests 

 - Unit test
 - Integration test

### CODE REVIEW

File: types/time.go [hunk 1]
```diff
@@ -649,6 +655,16 @@ func parseDatetime(str string, fsp int, isFloat bool) (Time, error) {
...
 		}
+		if l == 9 || l == 10 {
+			switch len(fracStr) {
+			case 0:
+				second = 0
+			case 1:
+				_, err = fmt.Sscanf(fracStr, "%1d", &second)
+			default:
+				_, err = fmt.Sscanf(fracStr[:2], "%2d", &second)
+			}
```
Comment (eurekaka) [inline types/time.go]: should we raise warning as MySQL does for the case below?
```
mysql> select cast("170102037.1122" as datetime);
+------------------------------------+
| cast("170102037.1122" as datetime) |
+------------------------------------+
| 2017-01-02 03:07:11                |
+------------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+------------------------------------------------------+
| Level   | Code | Message                                              |
+---------+------+------------------------------------------------------+
| Warning | 1292 | Truncated incorrect datetime value: '170102037.1122' |
+---------+------+------------------------------------------------------+
1 row in set (0.00 sec)
```

Comment (spongedu) [reply types/time.go]: Yes, indeed. I'll fix here

**Comment by zz-jason** [comment]
@spongedu We use go 1.11 now, maybe you can update `go` to make `fmt` happy.

**Comment by spongedu** [comment]
@zz-jason ok

**Comment by spongedu** [comment]
@zz-jason done here. PTAL

**Pull request merged**